src/prng/chacha.rs

```diff
@@ -274,14 +333,29 @@ mod test {
         let mut rng: ChaChaRng = SeedableRng::from_seed(seed);
         let mut results = [0u8; 32];
         rng.fill_bytes(&mut results);
-        // Same as first values in test_isaac_true_values as bytes in LE order
         let expected = [118, 184, 224, 173, 160, 241, 61, 144,
                         64, 93, 106, 229, 83, 134, 189, 40,
                         189, 210, 25, 184, 160, 141, 237, 26,
                         168, 54, 239, 204, 139, 119, 13, 199];
         assert_eq!(results, expected);
     }
 
+    #[test]
+    fn test_chacha_set_counter() {
+        // Test vector 5 from
+        // https://tools.ietf.org/html/draft-nir-cfrg-chacha20-poly1305-04
+        let seed = [0u8; 32];
+        let mut rng: ChaChaRng = SeedableRng::from_seed(seed);
+        rng.set_counter(0, 2u64.to_be());
```

Comment:
The user needs to call `.to_be()` on the indexes to get portable results? That is counter-intuitive for a counter. Are you sure it's correct, given that actually `set_counter` converts each `u64` to two `u32`s via bit-shift? I suspect it's not. (In fact, it's different from your example in the documentation.)

Think we should replace `set_counter` with something easier to use correctly? Would it be better to use four `u32`s?

Reply:
This comes from the test vector, which sets a 96-bit nonce to `00 00 00 00 00 00 00 00 00 00 00 02`. I am not sure I got it right with the `.to_be()` call. Probably should just not add that test, because in the comment I claimed we do not support using it for a nonce.

> Think we should replace set_counter with something easier to use correctly? Would it be better to use four u32s?

If we of think ChaCha in terms of an RNG and not a stream cipher, an `u128` counter is best. And two `u64`s an ok alternative. I don't think four `u32`s are handy to set a counter.

If we really want to set a nonce it should take a byte array, to match `from_seed`.

Just thought of another reason not to support setting a nonce: we still use it as part of a counter, and not as a constant value. So if you run the RNG long enough, eventually the counter will tick over and increment the nonce., producing 'incorrect' results.

Reply:
Then probably that test should use:
```rust
let counter = 2u64;
(counter >> 32) | (counter << 32)
```
 — but I don't understand why the test passes as-is (it sounds like `state[15]` should be set to `2u32`).

To be honest, I'm not really sure where this function would be useful, if at all. Perhaps we should also add:
```rust
fn set_nonce(&mut self, counter: u64, nonce: [u8; 8])
```

> So if you run the RNG long enough, eventually the counter will tick over and increment the nonce., producing 'incorrect' results.
The only correct behaviour would be to stop with an error, which conflicts with our primary usage (as a good RNG). If used to encrypt/decrypt a longer byte stream, I guess it has to be up to external code to handle this part correctly (e.g. split into chunks, or just continue encrypting with a note on compatibility).

Reply:
I opened an issue: https://github.com/dhardy/rand/issues/86

For now just try shifting the nonce as I suggested?

Reply:
>Then probably that test should use:
>```
>let counter = 2u64;
>(counter >> 32) | (counter << 32)
>```
>— but I don't understand why the test passes as-is (it sounds like `state[15]` should be set to `2u32`).

Your code misses the endian troubles... But my `2u64.to_be()` was wrong also because it does not get converted on BE. `rng.set_counter(0, 2u64 << 56);` works. But I am quite happy to see the test gone.

Reply:
A slice `[0u8, 0, 0, 2]` is the same as `0x0200_0000u32` in little-endian, the format we standardised upon inside the RNG. Together with the `u32`s that are in opposite order the value is `0x0200_0000_0000_0000u64`.